### ISSUE TITLE
False positive unreachable statement after exception raised within context block

### ISSUE THREAD

**Issue by DetachHead** [microsoft/pyright#6034]
from https://github.com/python/mypy/issues/8766:

> ```python
> import contextlib
> from typing import Iterator
> 
> @contextlib.contextmanager
> def context_func() -> Iterator[None]:
>     try:
>         yield
>     except Exception:
>         pass
> 
> def foo() -> None:
>     with context_func():
>         raise Exception
> 
>     print("xxx")  # mypy: Statement is unreachable
> ```

pyright has the same issue:

![image](https://github.com/microsoft/pyright/assets/57028336/d6a09a8c-bd16-44b6-b32a-be69b85eab79)

**Comment by erictraut** [comment]
The current behavior is as designed.

Some context managers swallow exceptions, others do not. Type checkers distinguish between them by looking at the annotated return type of the `__exit__` method for the context manager. If `__exit__` returns `Literal[True]` or `bool`, they assume that exceptions are swallowed. If the return type is annotated as `None`, `bool | None`, or `Literal[False]`. This standard was established by mypy, and pyright works consistently.

You can fix your code by creating a custom context manager rather than using the `contextlib.contextmanager` decorator.

**Comment by DetachHead** [comment]
> The current behavior is as designed.
> 
> Some context managers swallow exceptions, others do not. Type checkers distinguish between them by looking at the annotated return type of the `__exit__` method for the context manager. If `__exit__` returns `Literal[True]` or `bool`, they assume that exceptions are swallowed.

@erictraut that behavior is wrong though. an `__exit__` with the return type of `bool | None` can obviously return `True` which suppresses the error, making the code reachable at runtime:

```py
import contextlib
from types import TracebackType
from typing import Iterator, Literal

from typing_extensions import assert_never


class BoolOrNone(contextlib.AbstractContextManager[None]):
    def __exit__(
        self,
        __exc_type: type[BaseException] | None,
        __exc_value: BaseException | None,
        __traceback: TracebackType | None,
    ) -> bool | None:
        return True


with BoolOrNone():
    raise Exception
print(1)  # pyright/mypy thinks this is unreachable but it's reached at runtime
```

> This standard was established by mypy, and pyright works consistently.

where was this standard established exactly? the only place i can find where it was mentioned is https://github.com/python/mypy/issues/15158, which is an open issue stating that this behavior is a bug. you said on that issue it would be good to standardize the behavior, but i don't see why this obviously defective behavior should be the basis of this standard.

i think code being falsely marked as unreachable is a very serious problem in both pyright and mypy, unreachable code is not checked at all by either of the type checkers

**Comment by erictraut** [comment]
This was never standardized as part of a PEP. It was established by mypy and then adopted by typeshed, other stubs, etc. Pyright copied mypy's behavior here for consistency.

You can argue that the behavior is wrong in this particular case (with a return type of `bool | None`). If I had written the original mypy code, I would have probably treated this case differently. In fact, when I initially implemented this feature in pyright, I assumed `bool | None` would be treated like `bool` rather than like `None`, but I received bug reports indicating that pyright's behavior differed from mypy's.

Changing this behavior now would be very disruptive because there's significant bodies of typed code that rely on the current behavior. If mypy, pyright, or typeshed change the type checking behavior of `contextlib.contextmanager`, it will create a bunch of false positives in many code bases.

> i think code being falsely marked as unreachable is a very serious problem in both pyright and mypy

A type checker needs to know whether a context manager swallows exceptions or not. If it gets it wrong in either direction, false positives and/or false negatives will result. Getting this wrong in either direction is bad. Most context managers decorated with `contextlib.contextmanager` do not swallow exceptions, so regardless of how type checkers interpret the return result of `__exit__`, the typeshed maintainers would be wise to use a return type annotation that indicates that a context manager created with `contextlib.contextmanager` does not swallow exceptions. That will produce the correct type checker results most of the time. So the current definition in typeshed and the way that type checkers interpret that definition is correct the vast majority of the time.

Your code is the rare case where the implementation of a `@contextlib.contextmanager` does swallow exceptions, so the return type provided in typeshed will produce incorrect results. If you want to accurately reflect the behavior of your code to type checkers, you'll need to provide your own implementation of `@contextlib.contextmanager` that indicates that exceptions are swallowed.

**Comment by erictraut** [comment]
Here's what would happen if I were to make the change that you're proposing.

1. I'd change pyright's logic to interpret an `__exit__` return type of `bool | None` as "a context manager that swallows exceptions". This would (at least temporarily) make your code work the way you'd like. It would also make pyright behave inconsistently with mypy.
2. After releasing this change, pyright users would see hundreds or maybe even thousands of false positives in their code bases. Some subset of these users would file bugs. I'd point out "this change was intended, and pyright is doing what it's supposed to do; you'll need to talk to the maintainers of typeshed if you think that `@contextlib.contextmanager` should be changed".
3. The maintainers of typeshed would then change the return type of `_GeneratorContextManager.__exit__` to be `None` or `Literal[False]` to restore the older intended behavior and make it consistent between pyright and mypy.
4. I would then incorporate the updated typeshed stubs in the next release of pyright. This would once again restore the behavior of `@contextlib.contextmanager` that we see today. It would fix all of the new false positives, and it would once again "break" your code.

This would have resulted in massive churn across many code bases. It would have inconvenienced me, the typeshed maintainers, and hundreds or thousands of pyright users. And in the end, your code would still not work the way you hoped. So such a change doesn't make sense.

If you would like to champion a formal spec for the `__exit__` return type, I welcome you to do so.
If the there is consensus in the typing community that `bool | None` should be interpreted as "a context manager that swallows exceptions", then I would happily make the change in pyright, but I would coordinate the change with the maintainers of mypy and typeshed such that we all make the change at the same time and that the behavior of `@contextlib.contextmanager` does not change.

**Comment by DetachHead** [comment]
> 1. I'd change pyright's logic to interpret an `__exit__` return type of `bool | None` as "a context manager that swallows exceptions". This would (at least temporarily) make your code work the way you'd like. It would also make pyright behave inconsistently with mypy.
> 
> 2. After releasing this change, pyright users would see hundreds or maybe even thousands of false positives in their code bases. Some subset of these users would file bugs.

i don't think it would be that disruptive of a change. `bool | None` doesn't mean a "a context manager that swallows exceptions", it means "a context manager that *could* swallows exceptions". the code would be marked as reachable, that doesn't mean it *will* be reached, only that it *could* be reached.

for that reason, i feel like code being falsely marked as unreachable is far worse than code being falsely marked as reachable.

> The maintainers of typeshed would then change the return type of `_GeneratorContextManager.__exit__` to be `None` or `Literal[False]` to restore the older intended behavior and make it consistent between pyright and mypy.

i don't see why they would. the return type of `_GeneratorContextManager.__exit__` is correct, because `_GeneratorContextManager` could either suppress or not suppress the error. if they were to make a change, it would probably be making the `__exit__` return type generic so it can be specified on individual context managers. i'd be happy to try to make such a change

i just think type checkers should follow what the types say, instead of introducing strange special cases on random things just because another type checker does the same thing (which i'm still convinced wasn't even intentional in the first place. is there any discussion/documentation anywhere indicating otherwise?)

**Comment by erictraut** [comment]
Despite what you may think, this would be very disruptive. I don't plan to make any such change in pyright at this time for the reasons I explained above.

As I said, if you would like to champion a formal spec for the `__exit__` return type, I would welcome that effort. Formalization of typing standards is good in general, and I'd prefer if this were part of the official typing spec rather than an informal convention started by mypy.

If, as part of your draft PEP, you would like to propose that `@contextlib.contextmanager` should be annotated such that type checkers treat all such context managers as swallowing exceptions, you can do so, but I think you'll get strong pushback on that point during the PEP review process. As I mentioned above, the vast majority of functions decorated with `@contextlib.contextmanager` do not swallow exceptions, so it would be surprising to most users if it implied otherwise. 

Perhaps you could propose to add a new symbol to `contextlib` that has the same runtime implementation as `contextmanager` but is annotated differently such that it can be used in cases where the decorated function's implementation includes a `try`/`except` statement.
